controller_tuner_nets.py

```diff
@@ -0,0 +1,117 @@
+            nn.Linear(4 + 4 + 1, 16),
+            nn.Tanh(),
+            nn.Linear(16, 16),
+            nn.Tanh(),
+            nn.Linear(16, 16),
+            nn.Tanh(),
+            nn.Linear(16, 4),
+            nn.Softplus()
+        )
+
+    def forward(self, x):
+        x = self.controller(x)
+        return torch.exp(x)
+
+
+class ControllerExternalDerivativesV1(ControllerV1):
+    # TODO: universal Controller for all external controllers
+    def __init__(self):
+        super().__init__()
+        self.controller[0] = nn.Linear(4, 16)
+        self.controller[-1] = nn.Tanh()
+        # TODO: remove Dropout when apply -- add controller.eval() everywhere
+        # TODO: check
+
+    def forward(self, x):
+        return self.controller(x) * 50  # TODO: limit values with less crutches
+
+
+class TunerCoordinatesV1(nn.Module):
+    def __init__(self, ar=3):
+        super(TunerCoordinatesV1, self).__init__()
+        self.ar = ar
+        self.tuner = nn.Sequential(
+            nn.Linear(4 * ar, 16),
+            nn.Tanh(),
+            nn.Linear(16, 16),
+            nn.Tanh(),
+            nn.Linear(16, 16),
+            nn.Tanh(),
+            nn.Linear(16, 4),
+            nn.Tanh()
+        )
+
+    def forward(self, x):
+        # TODO: x2, y2 += 1
+        return self.tuner(x) * 2
+
+
+class TunerAnglesV1(nn.Module):
+    def __init__(self, ar=3, dropout=0.1):
+        super(TunerAnglesV1, self).__init__()
+        self.ar = ar
+        self.tuner = nn.Sequential(
+            nn.Linear(2 * ar, 16),
+            nn.Tanh(),
+            nn.Dropout(dropout),
+            nn.Linear(16, 16),
+            nn.Tanh(),
+            nn.Dropout(dropout),
+            nn.Linear(16, 16),
+            nn.Tanh(),
+            nn.Dropout(dropout),
+            nn.Linear(16, 2),
+            nn.Tanh()
+        )
+        # self._i = 0
+
+    # TODO: add length dependency
+    # TODO: пофиксить резкий переход в 0 на 360
+    # TODO: faster using torch stuff instead of numpy?
+    @staticmethod
+    def _coordinates2angles(coords):
+        coords = coords.numpy()
+        angles = np.stack([
+            np.angle(coords[0] + 1j * coords[1]),
+            np.angle(coords[2] - coords[0] + 1j * (coords[3] - coords[1]))
+        ]).T
+        return angles
+
+    def _coordinates2angles_ar(self, coords):
+        angles = []
+        for i in range(self.ar):
+            ar_coords = coords[:, i]
+            ang = self._coordinates2angles(ar_coords)
+            angles.append(ang)
+        angles = np.stack(angles)
+        angles = angles.transpose((2, 0, 1))
+
+        return angles
+
+    @staticmethod
+    def _angles2coordinates(angles):
+        x1 = torch.cos(angles[0])
+        y1 = torch.sin(angles[0])
+        x2 = torch.cos(angles[1]) + x1
+        y2 = torch.sin(angles[1]) + y1
+        return torch.stack([x1, y1, x2, y2])
+
+    def forward(self, x, a=1.):
+        angles = self._coordinates2angles_ar(x)
```

Comment:
Where and how this method is used?

`self._coordinates2angles_ar` turns coordinates into angles, but `DoublePendulumApproxDiffEq` outputs already angles. So there is no need to preprocess coordinates into angles.

Reply:
No, it seems that ```DoublePendulumApproxDiffEq``` returns derivatives. But yes, ```odeint(DoublePendulumApproxDiffEq)``` returns angles. For me, it is quite more convenient to work with coordinates, because:

1. in general NN Tuner adjust coordinates and tuning coordinates through angles is just one of the options
2. it is quite easier to test and debug 
3. it could be a good idea to get rid of preprocessing coordinates into angles in the production if exactly this approach will be used